Xiaow TimeEntryForm refactor

wang9hu opened the pull request:
# Description
This is the refactor of TimeEntryForm, TimeLog and TimeEntry along with related files. 

## Related PRS (if any):
To test this frontend PR you need to checkout the #[668](https://github.com/OneCommunityGlobal/HGNRest/pull/668#issue-2054757181) backend PR.
…

## Main changes explained:
- Restructure format of files, favoring destructure assignment, redux state management, and code categorization arranged in a centralized manner
- Add WBS and task names in the project/task selection for both TimeEntryForm and TimeLog
- Add wbs in state.userProject reducer for better data handling
- Add wbsId and taskId to timeentry data for better data displaying
- Refactor event handler logic
- Disable outdated test files of these component, will overhaul those in the future
…

## How to test:
1. check into current branch
2. check into #[668](https://github.com/OneCommunityGlobal/HGNRest/pull/668#issue-2054757181) backend PR.
3. do `npm install` and `npm run start:local` to run this PR locally
4. Clear site data/cache
5. log as admin/vol user
6. Test the five use cases of TimeEntryForm below, try to break it.

## Screenshots or videos of changes:

## Note:
- userProfile, userProjects and userTask reducers are solely for storing data displaying on page (not for logged in user, this is important). 

- There are five use cases of TimeEntryForm component, and there are two boolean values passed in props that affect the behavior of TimeEntryForm: fromTimer (whether it is opened from Timer), edit(whether it is opened from editting an old timeentry).
   1. Account Owner (logged in user, not owner role account) user logs new time entry from Timer to Account Owner
   2. Account Owner adds time entry from 'Add Intangible Time Entry' button to Account Owner
   3. Account Owner edits same day time entry of Account user
   4. Account with permission (like owner or admin) adds time entry from 'Add Intangible Time Entry for XXX' button to XXX user
   5. Account with permission edits existing time entry of other user
 
- The recommended approach for these 5 cases are:
  1. Login as any user (better to use volunteer account, since it has the least permissions), start timer and log time, check and see if there is any issue in the process of logging time, and if you could see the newly added time entry on the Time Log page.
  2. Login as any user, use `Add Intangible Time Entry` button from Time Log page to add intangible time entry, check for issues and the newly added time entry on Time Log page. 
  3. Login as any user, edit the time entry that was logged in the same day, check for issues and the new changes on that time entry. For users with limited edit times (like volunteer user), check if you could see the warning modal for 5 edit times limits. Make sure other time entries that are logged in prior days don't have the edit button shown on that entry.
  4. Login as owner/admin account, open the Time Log page of other account (not logged in account, you would see a yellow banner say `You are currently viewing the dashboard/Timelog for XXX`), use the `Add Intangible Time Entry` button from other account's page to add intangible time entry for other account, check for issues and the newly added time entry on the other account's Time Log page (you need to log in that account and see from there).
  5. Login as owner/admin account, open the Time Log page of other account, edit the time entry from other account's page, check for issues and the new changes on that time entry.

wang9hu committed:
move userProfile action creators from constants dir to actions dir, make sure any related files are changed accordingly

wang9hu committed:
make Image.test.jsx end of line LF

wang9hu committed:
Merge branch 'development' into xiaow_TimeEntryForm_refactor

wang9hu committed:
Merge branch 'development' into xiaow_TimeEntryForm_refactor

wang9hu committed:
Merge branch 'development' into xiaow_TimeEntryForm_refactor

wang9hu committed:
Refactor of TimeEntryForm along with other components

wang9hu committed:
Merge branch 'development' into xiaow_TimeEntryForm_refactor

wang9hu committed:
finish TimeEntryFrom, TimeLog, TimeEntry refactor

wang9hu committed:
small fix on clearForm

wang9hu committed:
Merge branch 'development' into xiaow_TimeEntryForm_refactor

wang9hu committed:
fix lint issue in WeeklySummay

wang9hu committed:
Merge branch 'xiaow_TimeEntryForm_refactor' of https://github.com/OneCommunityGlobal/HighestGoodNetworkApp into xiaow_TimeEntryForm_refactor

wang9hu committed:
small bug fix, disable outdated test files

wang9hu committed:
Merge branch 'development' into xiaow_TimeEntryForm_refactor

wang9hu committed:
fix popupbar test file

ChengyanOo commented:
> # Description
> This is the refactor of TimeEntryForm, TimeLog and TimeEntry along with related files.
> 
> ## Related PRS (if any):
> To test this frontend PR you need to checkout the #[668](https://github.com/OneCommunityGlobal/HGNRest/pull/668#issue-2054757181) backend PR. …
> 
> ## Main changes explained:
> * Restructure format of files, favoring destructure assignment, redux state management, and code categorization arranged in a centralized manner
> * Add WBS and task names in the project/task selection for both TimeEntryForm and TimeLog
> * Add wbs in state.userProject reducer for better data handling
> * Add wbsId and taskId to timeentry data for better data displaying
> * Refactor event handler logic
> * Disable outdated test files of these component, will overhaul those in the future
>   …
> 
> ## How to test:
> 1. check into current branch
> 2. check into #[668](https://github.com/OneCommunityGlobal/HGNRest/pull/668#issue-2054757181) backend PR.
> 3. do `npm install` and `npm run start:local` to run this PR locally
> 4. Clear site data/cache
> 5. log as admin/vol user
> 6. Test the five use cases of TimeEntryForm below, try to break it.
> 
> ## Screenshots or videos of changes:
> ## Note:
> * userProfile, userProjects and userTask reducers are solely for storing data displaying on page (not logged in user, this is important).
> * There are five use cases of TimeEntryForm component
> 
> 1. Account Owner (not owner account) user logs new time entry from Timer to Account Owner: edit is false, fromTimer is true, isTangible is true;
> 2. Account Owner adds time entry from 'Add Intangible Time Entry' button to Account Owner: edit is false, fromTimer is false, isTangible depends;
> 3. Account Owner with permission adds time entry from 'Add Intangible Time Entry for XXX' button to XXX user: edit is false, fromTimer is false, isTangible depends;
> 4. Account Owner edits same day time entry of Account user: edit is true, fromTimer is false, isTangble depends;
> 5. Account Owner with permission edits existing time entry of other user: edit is true, fromTimer is false, isTangble depends;

Hi @wang9hu,

I'm in the process of testing the feature and would appreciate some additional guidance. Could you clarify the meaning of "edit is false, fromTimer is true, isTangible is true"? Specifically, I need to understand how these boolean values impact the feature's functionality. Also, could you advise on the best method for verifying these values? For instance, is inserting console logs in the TimeEntry.jsx file the recommended approach, or is there another preferred method for this verification?

wang9hu commented:
> I'm in the process of testing the feature and would appreciate some additional guidance. Could you clarify the meaning of "edit is false, fromTimer is true, isTangible is true"? Specifically, I need to understand how these boolean values impact the feature's functionality. Also, could you advise on the best method for verifying these values? For instance, is inserting console logs in the TimeEntry.jsx file the recommended approach, or is there another preferred method for this verification?

Hi @ChengyanOo, thank you for your suggestions on this! I updated the description a little and hopefully that will clear up some confusions. I removed the `isTangble` since it is less impactable comparing to the other two boolean values.

ChengyanOo commented:
> > I'm in the process of testing the feature and would appreciate some additional guidance. Could you clarify the meaning of "edit is false, fromTimer is true, isTangible is true"? Specifically, I need to understand how these boolean values impact the feature's functionality. Also, could you advise on the best method for verifying these values? For instance, is inserting console logs in the TimeEntry.jsx file the recommended approach, or is there another preferred method for this verification?
> 
> Hi @ChengyanOo, thank you for your suggestions on this! I updated the description a little and hopefully that will clear up some confusions. I removed the `isTangble` since it is less impactable comparing to the other two boolean values.

@wang9hu 
Thank you for your explanation!

Alforoan reviewed (changes requested):
Everything works as intended except for editing (both when volunteer tries to edit and when the owner tries to edit the volunteer's.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/118868270/47655158-418d-4896-b750-dab50dcb2c0a

ChengyanOo commented:
@Alforoan 

Sorry that I accidentally re-requested a review for your PR review, didn't meant to do it

ChengyanOo commented on src/actions/userProfile.js:
Is it missing error handling near line 31-34?

wang9hu replied:
Error handling added in the latest commit.

ChengyanOo commented on src/actions/index.js:
Hi there! I noticed there's a section of code that's been commented out. Could you please add a brief comment explaining the reason for this? This would really help other developers understand the context behind this change. Also, if you're confident that this code won't be needed in the future, consider removing it to keep our codebase clean and tidy. Thanks!

wang9hu replied:
That action creator is for one reducer that is not included in our current store. It is only connected to one file that has only one reducer, which is not being used anywhere else, either.  Also it hasn't been changed in 4 years. So I think it is safe to just delete them and if we need similar functionality, it won't be hard to add it later.

ChengyanOo commented on src/actions/userProjects.js:
Is it missing error handling near line 23?

wang9hu replied:
Error handling added in the latest commit.

ChengyanOo reviewed (changes requested):
Hi @wang9hu,

Great job on the refactoring of multiple React components! I've conducted tests on the features you implemented, specifically in the scenarios involving volunteer and admin roles, as you suggested. However, I encountered the same issue as @Alforoan.

Below, I've attached my complete server log and a video walkthrough of these features for your reference:

- **Video Walkthrough**: [Feature Test Video](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/69735000/7f778581-d4d3-48dc-8a5c-d51f1dec2051)
- **Log File**: [HGNRest_pullReq_668_Dec29.log](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/files/13795234/HGNRest_pullReq_668_Dec29.log)

bienzguoa reviewed (commented):
There is something wrong with the "Edit" function( for both volunteers and admins). Others work as intended!
Logged as vol:
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/92126559/c14c34bb-d8b4-4a1f-b485-d8366792e420
Logged as admin:
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/92126559/42ded897-7ead-4902-a491-cdcbbb863664

wang9hu committed:
add error handlings for get user task and get user project requests, delete unused file and action creator, fix a bug in TimeEntry

wang9hu commented:
Thank you all for the prompt feedbacks, @Alforoan @ChengyanOo and @bienzguoa!

I have fixed the issue and made some changes according to those feedbacks. Please take another look at it and let me know if there is more bugs.

one-community commented:
Thank you all, merging!
